src/main/scala/tech/cryptonomic/conseil/tezos/TezosPlatformDiscoveryOperations.scala

```diff
@@ -209,26 +231,57 @@ class TezosPlatformDiscoveryOperations(
-      val highCardinalityValidationResult = if (!isLowCardinality(attrOpt.flatMap(_.cardinality))) Some(HighCardinalityAttribute(column)) else None
-      val validationErrors = List(invalidDataTypeValidationResult, highCardinalityValidationResult).flatten
-
-      val res = Either.cond(
-        test = validationErrors.isEmpty,
-        right = attrOpt.map(attr => makeAttributesQuery(tableName, attr.name, withFilter)).toList.sequence.map(_.flatten),
-        left = Future.successful(validationErrors)
-      )
+  def listAttributeValues(tableName: String, column: String, withFilter: Option[String] = None, attributesCacheConfig: Option[AttributeCacheConfiguration] = None): Future[Either[List[AttributesValidationError], List[String]]] = {
+    getTableAttributesWithoutUpdatingCache(tableName) map (_.flatMap(_.find(_.name == column))) flatMap  { attrOpt  =>
+      val res = (attributesCacheConfig, withFilter) match {
+        case (Some(AttributeCacheConfiguration(cached, minMatchLength, maxResultLength)), Some(attributeFilter)) if cached  =>
+          Either.cond(
+            test = attributeFilter.length >= minMatchLength,
+            right = getAttributeValuesFromCache(tableName, column, attributeFilter, maxResultLength),
+            left = Future.successful(List(InvalidAttributeFilterLength(column, minMatchLength)))
+          )
+
+        case _ =>
+          val invalidDataTypeValidationResult = if (!attrOpt.exists(attr => canQueryType(attr.dataType))) Some(InvalidAttributeDataType(column)) else None
+          val highCardinalityValidationResult = if (!isLowCardinality(attrOpt.flatMap(_.cardinality))) Some(HighCardinalityAttribute(column)) else None
+          val validationErrors = List(invalidDataTypeValidationResult, highCardinalityValidationResult).flatten
+          Either.cond(
+            test = validationErrors.isEmpty,
+            right = attrOpt.map(attr => makeAttributesQuery(tableName, attr.name, withFilter)).toList.sequence.map(_.flatten),
+            left = Future.successful(validationErrors)
+          )
+      }
       res.bisequence
     }
-    result.flatten
   }
 
+  /** Gets attribute values from cache and updates them if necessary */
+  private def getAttributeValuesFromCache(tableName: String, columnName: String, attributeFilter: String, maxResultLength: Int): Future[List[String]] = {
+    attributeValuesCache.read.flatMap { cache =>
+      cache(makeKey(tableName, columnName)) match {
+        case (last, radixTree) if last + cacheTTL.toMillis > now =>
+          IO.pure(radixTree.filterPrefix(attributeFilter.toLowerCase).values.take(maxResultLength).toList)
+        case (_, oldRadixTree) =>
+          for {
+            cache <- attributeValuesCache.take
+            _ <- attributeValuesCache.put(cache.updated(makeKey(tableName, columnName), (now, oldRadixTree)))
```

Comment:
No need to put something for other threads to see, here, as I see it.
The `take` call will block any other read from this method, and it's ok.
Once we have taken the `cache`, we only need to verify once more if the value for the `key` is still stale
```
cache(key)._1 + cacheTTL.toMillis > now?
```
if not, it will mean another thread updated the cache on this key before we did, so the new value is already fine and we can immediately put back the cache into the MVar.
If it's expired, then we first need to actually update the tree and put it in the cache, before updating the MVar

The current solution takes and puts back twice, which makes reasoning on concurrency even harder than before... at the risk of introducing some bug

Reply:
I was thinking of the optimized solution - if we would `take` something from the cache we would need to wait on all threads till the first computation finishes, but this way we will still use old data, and when first computation finishes it will immediately update the cache, so only one call will wait for the database to complete query(which by the way is really heavy).

Reply:
Assume that 2 threads find a stale cache and race to update it, they're both al line `268`
- t1 takes the cache
- t2 takes but needs to wait for a callback
- t1 puts the old value with a new update-time
- t2 takes the cache and puts back again the old value with yet another new update-time
- t1 computes the new query
- t2 computes the same, on another thread... increasing contention on the db too (as a secondary side-effect)
- both take and put back the newly computed value in the cache, (order is irrelevant)

It's not so critical a scenario, but it shows that we've not avoided the original problem of making multiple updates (I suppose the same exact thing can involve more than 2 threads)

Given that, I'd argue that we would better leave the code as it was before... it was simpler to reason about, and it's less amenable to unforseen contention and racing conditions.
We're talking about a rare case for sure, but it's all about making robust design decision, to take those pesky cases seriously.

Your thoughts?

Reply:
I understand your concern, but in previous case we had really big time window for this race condition. Updating distinct values of the attributes may take over 30 seconds, so if we don't put back cache, every single request during those 30 seconds will cause another heavy DB call to be queued. 

In this case only requests which come before the first put are going to be affected which is significantly smaller window, I would guess few milliseconds.

Reply:
From what I learned, there's no real way to estimate with certainty how much time will pass between execution of two instructions in the code. While the general value will certainly reflect what you say, the tail can be unexpectedly long, due to many possible reasons, a long context switch, the compiler having re-ordered the code for JIT optimization, a GC pause. I never make assumptions here.

I see no great reasons to be concerned apparently, I simply hope we'll never find ourselves in a position to need to investigate any concurrency trouble here, in the long run.

I simply tend to lean on the safe side if I can, but I leave this up to you.

Reply:
I think for now we can leave it as it is. I think in the future we could think of different way of handling attribute values cache, possibly using triggers from postgres(but from what I've seen Slick does not support that), and it would change architecture a bit.